### PR TITLE
[Model CI] Full client support for current functionality

### PR DESCRIPTION
**Pull request by phil-scale**
Adds Python client support for the following
- Creating unit tests and unit test metrics
- Evaluating a model on unit tests
- Getting the metrics associated with a unit test
- Getting the evaluation history of a unit test

This PR also introduces a `modelci` folder and a `ModelCIClient` that inherits from the base `NucleusClient`. This is an effort to make the relationship between Nucleus and Model CI directed, where Model CI currently depends on Nucleus but Nucleus does not need to depend on Model CI. The separate folder and client should also allow for easier refactoring in the future.

[[sc-262746]](https://app.shortcut.com/scaleai/story/262746)
[[sc-269906]](https://app.shortcut.com/scaleai/story/269906)

### CODE REVIEW

File: nucleus/modelci/utils.py
Comment (yixu34) [inline nucleus/modelci/utils.py:8]: `NaN` has a [special meaning in the definition of floats](https://en.wikipedia.org/wiki/NaN), e.g. when you do `0 / 0`. I think we want to surface some other kind of error string if someone passes in something that wasn't even a valid float to begin with.

File: tests/modelci/test_unit_test.py
Comment (yixu34) [inline tests/modelci/test_unit_test.py:64]: Might want to be explicit here because python treats zeroes as false-y.

File: tests/modelci/test_unit_test.py
Comment (yixu34) [inline tests/modelci/test_unit_test.py:65]: Some of the other fields might be auto-generated, or it'd be inappropriate to peek into implementation details, but in this case, it should be OK to check for the `GREATER_THAN` that you passed in, right?

File: tests/modelci/test_unit_test.py
Comment (yixu34) [inline tests/modelci/test_unit_test.py:79]: Wonder how the Nucleus folks feel about creating a util for this `str(uuid.uuid4())[-N_UUID_CHARACTERS:]` pattern that keeps coming up.

File: nucleus/modelci/unit_test_evaluation.py
Comment (yixu34) [inline nucleus/modelci/unit_test_evaluation.py:11]: Curious why we need to write this as a single-item tuple 🤔 

Seems neat that we're treating it as both `str` and `Enum` though.

**Review by yixu34 — COMMENTED** [review]
A concrete class inheriting from another is generally a smell. Here's one article that talk about it: https://www.rangakrish.com/index.php/2016/08/07/is-deriving-from-a-concrete-class-bad/

One way to think of it is through the Liskov substitution principle. Seems slightly weird IMO to say that "a ModelCI client is-a Nucleus client". It also means that we're now surfacing two clients in the client library. Like I get the idea at the moment, but things might diverge in unexpected ways down the line.

But, seeing as how generally we're invested in ModelCI leveraging Nucleus, I would almost think of ModelCI as a mixin/expansion pack to the Nucleus client library. This leverages composition instead of inheritance. So it'd look something like:

```python3
class ModelCIClientMixin:
    <functionality>
```
https://docs.github.com/github/writing-on-github/getting-started-with-writing-and-formatting-on-github/basic-writing-and-formatting-syntax
```python3
class NucleusClient(ModelCIClientMixin):
```

Or if you want to do it more dynamically/explicitly, you could pass it through a constructor:
```python3
class NucleusClient:
    def __init__(self, addons=...)
```

File: nucleus/modelci/unit_test_evaluation.py
Comment (phil-scale) [reply nucleus/modelci/unit_test_evaluation.py:11]: Not sure, this is copied from code that @sasha-scale previously wrote.

File: nucleus/modelci/utils.py
Comment (phil-scale) [reply nucleus/modelci/utils.py:8]: I do think we should still fail gracefully, but maybe I can just log an error as well?

Comment (yixu34) [reply nucleus/modelci/utils.py:8]: Not sure if we need to log stuff in user code, that should be on them (it does bring up the question of what kind of API we want to provide for user-defined logging, if any). Though if we're provided any pre-built code, then I think it can make sense to log ourselves.

We should still be able to fail gracefully while still being type-correct. I think this stems from `UnitTestItemEvaluation` having `result` being a `float`, when IMO it should be an `Optional[float]`. The db table should correspondingly have the column be `nullable`, to signify an exception in the code (maybe you'd need an extra `nullable` column for an error code/string). Overloading `NaN` to mean both an exception and `0 / 0` could be a confusing edge case.

Comment (phil-scale) [reply nucleus/modelci/utils.py:8]: Ah ok `Optional[float]` makes sense - sounds good.

File: tests/modelci/test_unit_test.py
Comment (phil-scale) [reply tests/modelci/test_unit_test.py]: Yup thanks!

File: tests/modelci/test_unit_test.py
Comment (phil-scale) [reply tests/modelci/test_unit_test.py]: Yup, that makes sense, thanks!

File: tests/modelci/test_unit_test.py
Comment (phil-scale) [reply tests/modelci/test_unit_test.py]: Added a helper method for this.k

File: nucleus/modelci/unit_test_evaluation.py
Comment (phil-scale) [reply nucleus/modelci/unit_test_evaluation.py:11]: I just changed them to strings, which I think should be ok.

**Comment by yixu34** [comment]
Ok to recap our offline discussion today around my [earlier comment on inheritance and whatnot](https://github.com/scaleapi/nucleus-python-client/pull/149#pullrequestreview-802104064):
* Mixin inheritance looks weird. So instead, we can use proper composition (i.e. `NucleusClient` takes in a `ModelCIClient` in the `NucleusClient` constructor).
* Alternatively: the main reason why we tried inheritance in the first place was to reuse the low-level request code. That can be refactored into a "neutral" module that is used by both `NucleusClient` and `ModelCIClient`. Since we're just using IDs between the two systems (i.e. they don't use the same client dataclasses), it's safe to do this.
* However, the Sia team wrote up some thoughts on the design of the Nucleus client library, wherein they propose that we move more in the direction of proper client data classes, which might make the above point trickier: https://scaleapi.slack.com/archives/G019J8773HA/p1635766488004600

For now, we can go with the current class design, but we should start a discussion asap.

File: nucleus/modelci/__init__.py
Comment (sasha-scale) [inline nucleus/modelci/__init__.py:189]:
```suggestion
            client = nm.ModelCIClient("YOUR_SCALE_API_KEY")
```

**Review by sasha-scale — CHANGES_REQUESTED** [review]
Can you add pytest coverage for client.list_unit_tests?

Before checking this in, take a look at the autogenerated docs by running:

```
poetry update
cd docs
poetry run sphinx-autobuild . ./_build/html --watch ../nucleus
```

There's some funkiness going on - most notably that due to the class inheritance, all nucleus client documentation is dunked into the model_ci page.  There's probably some sphinx setting to diable this behavior that we should take a look at, I recommend reaching out to Drake.

In general with code snippets in the documentation, it's important to check that they run successfully before shipping - we've gotten into trouble with example code that doesn't actually run in the past (its an easy mistake to make 😓)

**Comment by phil-scale** [comment]
Thanks for all the helpful comments! The tests are failing because the evaluation function `IOU` is inaccessible to the test account, so I think I'll first implement the fix for eval functions before continuing with this PR.

**Comment by phil-scale** [comment]
Hey @sasha-scale and @yixu34 - I went with making `ModelCI` an "extension" to the original `NucleusClient`. I'm not sure that this is the right approach to go with, but I think it's fairly flexible and easy to change at this point. Would love to get your feedback of this interface, using `nucleus_client.model_ci.create_unit_test` and so on.

File: nucleus/connection.py
Comment (yixu34) [inline nucleus/connection.py:42]: I would probably just import some retry library to help with this. I see that we use `tenacity` in `models` FWIW. Doing an exponential backoff with base 3 instead of base 2 is a bit quirky 😛

**Review by yixu34 — COMMENTED** [review]
FWIW I think `nucleus_client.modelci.create_unit_test` is OK as an interface, especially if we're thinking of Nucleus as the overall platform that users interact with. Curious what @sasha-scale thinks.

Also, we touched upon this during sprint planning, but this PR is getting a bit "horizontal" in the sense that we're lumping a bit too many changes in here that coincidentally happen to touch the client library. I'd say let's freeze the scope right now and try and push this through. My main ask is to consider a library for retries; will def to @sasha-scale for the rest.

File: nucleus/connection.py
Comment (phil-scale) [reply nucleus/connection.py:42]: I'll add a comment for this as a TODO. I don't want to lump that into this PR - this code was directly copied from the Nucleus base python client, so I think that we could introduce a change there in an individual PR.

Comment (phil-scale) [reply nucleus/connection.py:43]: Created this ticket to track: https://app.shortcut.com/scaleai/story/281630/use-external-library-for-http-retry-logic

File: nucleus/__init__.py
Comment (phil-scale) [reply nucleus/__init__.py]: Fixed, thanks!

File: nucleus/constants.py
Comment (phil-scale) [reply nucleus/constants.py]: Yup, removed - thanks!

File: nucleus/modelci/unit_test.py
Comment (phil-scale) [reply nucleus/modelci/unit_test.py]: Updated to just `nucleus.NucleusClient`, and we can use `nucleus.modelci.create_unit_test`

File: nucleus/modelci/__init__.py
Comment (phil-scale) [reply nucleus/modelci/__init__.py]: Fixed, thanks!

File: nucleus/modelci/__init__.py
Comment (phil-scale) [reply nucleus/modelci/__init__.py]: Same here

File: nucleus/modelci/__init__.py
Comment (phil-scale) [reply nucleus/modelci/__init__.py]: Ditto

**Comment by phil-scale** [comment]
Final pytest is failing because the endpoint for evaluation takes too long - I don't think that's a Python client problem, so I'll leave this test file as a separate PR: https://github.com/scaleapi/nucleus-python-client/pull/154

File: nucleus/modelci/__init__.py
Comment (sasha-scale) [inline nucleus/modelci/__init__.py:224]: Documentation is slightly incorrect here, if the unit_test_names are omitted, the model will be evaluated on all unit tests.  The array of names is a subsampling method, if that makes sense

**Review by sasha-scale — APPROVED** [review]
Some suggestions for improved usability:
- Improve string representations of classes to be consistent with other classes, here's an example:
https://github.com/scaleapi/nucleus-python-client/blob/2eb3aaf9fc1401febba682980dbcc14196713d86/nucleus/dataset.py#L69

Looks like `unit_test.get_eval_history()` doesn't work, should fix before merge

File: nucleus/modelci/__init__.py
Comment (phil-scale) [reply nucleus/modelci/__init__.py:224]: Hmm I think that no matter what, the model will only be evaluated on the specified unit tests, since `unit_test_names` is not optional.

In the case that it's the empty list, we would pass an empty list into the `TestEvaluator` constructor, and it looks to me like the subsampling is only checking if the input list is null, not if it's empty (see https://github.com/scaleapi/scaleapi/blob/d994d223ca552aee814f7c8fadbe26a6e6ee49bb/server/models/nucleus/modelci/MCUnitTest.ts#L209-L211)

Comment (phil-scale) [reply nucleus/modelci/__init__.py:224]: Opened up `node` just to confirm
```
> tests = []
[]
> if (tests) console.log("hello")
hello
```

File: nucleus/modelci/__init__.py
Comment (phil-scale) [reply nucleus/modelci/__init__.py:60]: Thanks!

File: nucleus/connection.py
Comment (phil-scale) [reply nucleus/connection.py:19]: Yeah credit to @yixu34 for this!

**Comment by phil-scale** [comment]
I just tried locally and `get_eval_history` works with me - it's `evaluate_model_on_unit_tests` that is problematic for me, but I think the problem is on the `scaleapi` side

```
>>> unit_tests = client.modelci.list_unit_tests()
>>> unit_tests[0].get_eval_history()
[]
>>> print(unit_tests[0])
UnitTest(unit_test_id='ut_c6ay39xmav9g048azy30', client=ModelCI(connection='Connection(api_key='scaleint_57d7743357384b9184e1030d16751880|615cc6e424ecf90020baffdd', endpoint='https://api.scale.com/v1/nucleus')')
>>>
```

File: tests/test_slice.py
Comment (phil-scale) [reply tests/test_slice.py:175]: <3

File: nucleus/modelci/__init__.py
Comment (phil-scale) [reply nucleus/modelci/__init__.py]: done, thanks!

File: nucleus/modelci/__init__.py
Comment (phil-scale) [reply nucleus/modelci/__init__.py]: done, thanks!

**Pull request merged**
